File: 2023/src/day1.rs
use regex::Regex;

#[aoc(day1, part1)]
pub fn part1(input: &str) -> u32 {
    input
        .lines()
        .map(|line| line.chars().filter_map(|c| c.to_digit(10)).collect())
        .map(|c: Vec<u32>| c.first().unwrap() * 10 + c.last().unwrap())
        .sum()
}

fn find(re: &Regex, line: &str) -> usize {
    match re.captures(line).unwrap().get(1).unwrap().as_str() {
        "one" => 1,
        "two" => 2,
        "three" => 3,
        "four" => 4,
        "five" => 5,
        "six" => 6,
        "seven" => 7,
        "eight" => 8,
        "nine" => 9,
        n => n.parse::<usize>().unwrap(),
    }
}

#[aoc(day1, part2)]
pub fn part2(input: &str) -> usize {
    let first = Regex::new(r".*?(\d|one|two|three|four|five|six|seven|eight|nine)").unwrap();
    let last = Regex::new(r".*(\d|one|two|three|four|five|six|seven|eight|nine)").unwrap();
    input
        .lines()
        .map(|line| find(&first, line) * 10 + find(&last, line))
        .sum()
}

File: 2023/python/day1.py
#!/usr/bin/env python3

from aoc import *


def part1(input: str):
    total = 0
    for c in comb(input, r"\d"):
        if c:
            total += int(c[0] + c[-1])
    return total


def toInt(input: str):
    match input:
        case "one":
            return "1"
        case "two":
            return "2"
        case "three":
            return "3"
        case "four":
            return "4"
        case "five":
            return "5"
        case "six":
            return "6"
        case "seven":
            return "7"
        case "eight":
            return "8"
        case "nine":
            return "9"
        case _:
            return input


def part2(input: str):
    total = 0
    exp = r"(\d|one|two|three|four|five|six|seven|eight|nine)"
    first = re.compile(exp)
    last = re.compile(".*" + exp)
    for line in input.splitlines():
        a = toInt(first.search(line).group(1))
        b = toInt(last.match(line).group(1))
        total += int(a + b)
    return total


if __name__ == "__main__":
    main()
